### PR TITLE
#189 DB Variance w/ Both-sides Absorption

### PR DESCRIPTION
**Pull request by xinhew0708**
This PR addresses both-sides absorption in design-based variance estimation. To calculate a DB variance with both-sides absorption for a teemod with absorption: `vcov_tee(teemod_abs, type = "DB0", const_effect = TRUE)`.

### CODE REVIEW

File: R/SandwichLayerVariance.R [hunk 1]
```diff
@@ -626,29 +626,27 @@ vcov_tee <- function(x, type = "CR0", cluster = NULL, ...) {
...
     a22inv <- sandwich::bread(x)
     meat <- do.call(sandwich::meatCL, args)
-
     vmat <- (1 / n) * a22inv %*% meat %*% a22inv
-    name <- paste0(var_names(x@StudySpecification, "t"), ".")
-    vmat <- as.matrix(vmat[name, name])
   }
   else {
+    if (length(x@moderator) > 0){
+      stop(paste("Design-based standard errors cannot be computed for teeMod",
```
Comment (benthestatistician) [inline R/SandwichLayerVariance.R:646]: Suggest "cannot be computed" -> "not supported"

File: R/SandwichLayerVariance.R [hunk 2]
```diff
@@ -1040,14 +1048,16 @@ vcov_tee <- function(x, type = "CR0", cluster = NULL, ...) {
   b_ind <- sapply(unique(stratum), function(val) as.integer(stratum == val))
 
   # nuisance parameters p
+  n <- length(stratum) # number of units in Q
+  ws <- if (is.null(stats::weights(x))) rep(1,n) else stats::weights(x)
   wb <- matrix(replicate(s, ws), ncol = s) * b_ind
-  p <- t(z_ind) %*% wb / (matrix(1, nrow = k, ncol = n) %*% wb)
+  p <- t(z_ind) %*% wb / (matrix(1, nrow = 2, ncol = n) %*% wb)
```
Comment (benthestatistician) [inline R/SandwichLayerVariance.R:1054]: Is 2 the no. of conditions?  (Seeing that it replaces "`k`".) If could you pls add a brief inline comment identifying it as such?

File: R/SandwichLayerVariance.R
Comment (benthestatistician) [inline R/SandwichLayerVariance.R:1]: Thanks for all the helpful inline comments that were added to this file!

**Review by benthestatistician — APPROVED** [review]

File: R/SandwichLayerVariance.R (hunk 1 above)
Comment (xinhew0708) [reply R/SandwichLayerVariance.R]: addressed in [e2b9fe0](https://github.com/benbhansen-stats/propertee/pull/213/commits/e2b9fe01bd867a458e53a8f473e8baea53661b42)

File: R/SandwichLayerVariance.R (hunk 2 above)
Comment (xinhew0708) [reply R/SandwichLayerVariance.R]: addressed in [e2b9fe0](https://github.com/benbhansen-stats/propertee/pull/213/commits/e2b9fe01bd867a458e53a8f473e8baea53661b42)

**Comment by benthestatistician** [comment]
I see nice updates of existing tests, as well as tests of the newfunction `block_center_residuals()`.  But I don't see the new functionality being tested directly -- i.e. no tests of `vcov_tee()`  that use arguments `type = "DB0"` and  `const_effect = TRUE`.  

We do need integration tests, at a minimum to show that that runs and produces output of the desired form, as well as tests of its constitutent operations.  Perhaps I'm missing something; if not could you add such integration tests @xinhew0708?

**Comment by xinhew0708** [comment]
Thanks for the suggestions, Ben!
* Integration tests added for `vcov_tee(..., type = "DB0", const_effect = TRUE)` and the mirror case with `const_effect = FALSE` in `test.SandwichLayerVariance.R`.
* Added basic `expect_*()` sanity checks to confirm both calls return valid variance–covariance matrices of the correct shape and that their values differ as expected.
* Added a few extra sanity checks for other types of tee models with absorption.

**Comment by benthestatistician** [comment]
Excellent, just was I was hoping for Xinhe.  If you have a chance to give a glance to this merge request and share any suggestions that may arise, @jwasserman2, please do. Otherwise I see this as ready to merge. 

There's something I'd like to ponder before merging, which is whether "DB0" is the best moniker for this calculation. Like HC0, it doesn't do anything specific to address degrees of freedom. In contrast to HC0, it's not clear that it should.  Without clustering or covariates but with larger or non-fine strata, it boils down to HC2, which suggests "DB2."  OTOH it doesn't bear much resemblance to CR2.

I'd be more comfortable with "DB0" as the official identifier for this calculation if "DB2" could be routed here also. I.e. we just add `.vcov_DB <- .vcov_DB0` somewhere, with a brief Roxygen stub so that it gets properly indexed. OK by you, @jwasserman2 and @josherrickson?

File: R/SandwichLayerVariance.R [hunk 3]
```diff
@@ -776,23 +780,29 @@ vcov_tee <- function(x, type = "CR0", cluster = NULL, ...) {
 .get_DB_covadj_meat <- function(x, ...) {
   agg <- .aggregate_to_cluster(x)
   data <- agg$data
-  bid <- data[, agg$block]  # block ids
-  zobs <- data[, agg$z] # observed zs
+  bid <- data[, agg$block] # block ids
+  zobs <- data[, agg$z] # treatment assignments
 
+  # number of covariates in the covariance model
   p <- ncol(stats::model.frame(x$model$`(offset)`@fitted_covariance_model))
```
Comment (jwasserman2) [inline R/SandwichLayerVariance.R:787]: I think you want `model.matrix()` here instead of `model.frame()`. `model.frame()` includes columns for weights and offsets

**Review by benthestatistician — APPROVED** [review]
- Please check for accuracy my small changes to internal documentation and warnings in commit 491d876.
- `.vcov_DB0()` and `.get_DB_covadj_se()` have a number of requirements enforced with code along lines of `if (!<requirement>) stop("models not meeting <requirement> not supported")`.  For purposes of future development, it would be helpful to have inline comments at these places indicating whether the requirement comes from the underlying theory, versus the software not (yet) being equipped to handle failure of the condition. 
- That doesn't need to happen before the merge of this PR. To me it looks ready to go now!

**Comment by xinhew0708** [comment]
Thank you for the comments, Ben. 

- Your changes in commit [491d876](https://github.com/benbhansen-stats/propertee/commit/491d876aaa72cab4bf566d74f44c523d483093bf) look good to me.
- I've added in-line comments clarifying requirement conditions in `.vcov_DB0()` and `.get_DB_covadj_se()`.
